### PR TITLE
RNAhybrid mde computation

### PR DESCRIPTION
**Pull request by sjanssen2**
RNAhybrid computes the p-value of the predicted target-miRNA duplex as (from https://rnajournal.cshlp.org/content/10/10/1507.full.pdf+html):
image
`pvalue = 1-exp(-exp(-(-normalised_energy-xi)/theta));`
https://github.com/jlab/RNAhybrid/blob/bdd121baa3c91b4e147921dafd3f6e80cebd9512/Original_ADPc/RNAhybrid-2.1.2/src/hybrid_core.c#L1721

e_n = `normalised_energy` is the free energy (`v2`) of the duplex divided by the log of the product of the length of **both** input sequences (`n` times `m`): `normalised_energy = v2/log(m*n);` = https://github.com/jlab/RNAhybrid/blob/bdd121baa3c91b4e147921dafd3f6e80cebd9512/Original_ADPc/RNAhybrid-2.1.2/src/hybrid_core.c#L1720

If xi and theta are not given via command line, they are computed as 
```
      mde = maximal_duplex_energy(query_sq);
      xi    =    xi_slope * mde +    xi_intercept;
      theta = theta_slope * mde + theta_intercept;
```
https://github.com/jlab/RNAhybrid/blob/bdd121baa3c91b4e147921dafd3f6e80cebd9512/Original_ADPc/RNAhybrid-2.1.2/src/rnahybrid.c#L383-L388

where `query_sq` is the miRNA input sequence, `xi_slope`, `xi_intercept`, `theta_slope` and `theta_intercept` are pre-defined for three organisms (fly, worm and human, command line parameter `-s` OR can be estimated with the program RNAcalibrate) in https://github.com/jlab/RNAhybrid/blob/bdd121baa3c91b4e147921dafd3f6e80cebd9512/Original_ADPc/RNAhybrid-2.1.2/src/globals.h#L41-L54

The `mde` is the maximal energy of the miRNA input perfectly paired in one single stem with its complement. This is computed via the new instance of this PR.

### CODE REVIEW

**Review by kmaibach — APPROVED** [review]
So, do I understand correctly, that the MDE instance is only needed - in the end - to compute the p-value for one miRNA-target-RNA pair? 
Is this supposed to happen every time RNAhybrid is executed or only when the user adds a certain parameter?

I believe the webversion of RNAhybrid also gives the option of selecting the aprrox. p-value but does not output the calculated value when you don't choose between the 3 pre-defined ones (human, fly, worm).

File: rnahybrid.gap [hunk 1]
```diff
@@ -429,9 +429,21 @@ grammar gra_rnahybrid uses sig_rnahybrid(axiom = hybrid) {
          # h;
 }
 
+/* This is a grammar to compute the maximal possible free energy for the miRNA sequence,
+   iff it hybridizes with itself, i.e. second input must be the complement of first!
```
[the remark this answers is not in the document]
Comment (sjanssen2) [reply rnahybrid.gap:433]: I wanted to use it as `if and only if`: https://en.wikipedia.org/wiki/If_and_only_if

File: rnahybrid.gap [hunk 2]
```diff
@@ -429,9 +429,21 @@ grammar gra_rnahybrid uses sig_rnahybrid(axiom = hybrid) {
          # h;
 }
 
+/* This is a grammar to compute the maximal possible free energy for the miRNA sequence,
```
[the remark this answers is not in the document]
Comment (sjanssen2) [reply rnahybrid.gap:432]: you are right, terminology is confusing. We do optimize but with negative sign, i.e. best is minimum. We should better call it "optimal duplex energy", but I am sticking to the RNAhybrid 2.0 / 1.0 code style

**Comment by sjanssen2** [comment]
> Is this supposed to happen every time RNAhybrid is executed or only when the user adds a certain parameter?

once per provided miRNA. I think you can call RNAhybrid on a multiple fasta file. More important to us: we intend to compute sub- or co-optimal or a pareto front of results. We want to compute p-values for all these results and need their individual free energy - but only once the optimal duplex energy - since the miRNA sequence will not change.

> I believe the webversion of RNAhybrid also gives the option of selecting the aprrox. p-value but does not output the calculated value when you don't choose between the 3 pre-defined ones (human, fly, worm).
xi and theta can be provided as user arguments or via alias like human3utr. This will become important for the CLI: https://github.com/jlab/RNAhybrid/issues/28

**Pull request merged**